sql/rowexec: implement vector search operators in the row engine

DrewKimball opened the pull request:
#### sql: make vecindex.Manager available through ServerConfig

This commit adds and populates a `VecIndexManager` field in
`execinfra.ServerConfig`. This will be used by vector search execution
operators in a later commit.

Epic: CRDB-42943

Release note: None

#### sql/rowexec: implement vector search operators

This commit implements the `VectorSearch` and `VectorMutationSearch`
operators in the row engine. This required adding two new methods to
the persistent store - `WrapTxn` to get a `vecstore.Txn` from an open
SQL transaction, and `QuantizeAndEncode` to quantize and encode vectors
for put operations.

Epic: CRDB-42943

Release note: None

DrewKimball committed:
sql: make vecindex.Manager available through ServerConfig

This commit adds and populates a `VecIndexManager` field in
`execinfra.ServerConfig`. This will be used by vector search execution
operators in a later commit.

Epic: CRDB-42943

Release note: None

DrewKimball committed:
sql/rowexec: implement vector search operators

This commit implements the `VectorSearch` and `VectorMutationSearch`
operators in the row engine. This required adding two new methods to
the persistent store - `WrapTxn` to get a `vecstore.Txn` from an open
SQL transaction, and `QuantizeAndEncode` to quantize and encode vectors
for put operations.

Epic: CRDB-42943

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141897)
<!-- Reviewable:end -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141897)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`pkg/sql/rowexec/vector_search.go` line 201 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-OJvKCTQCmBFMUZb1TUD:-OJvKCTQCmBFMUZb1TUE:b-y9sleg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9017cf6a5adfd443d38af38fc8865b66b474d337/pkg/sql/rowexec/vector_search.go#L201)):*
> ```Go
> 	// Continue reading data until there's none left, or we've finished populating
> 	// the data for all the requested composite-typed columns.
> 	for len(compositeValBytes) > 0 {
> ```

I'm surprised there isn't a pre-existing utility function somewhere to do this. Surely it's being done elsewhere?
___
*[`pkg/sql/rowexec/vector_search.go` line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-OJvKjmu1Kl0IyPg6UhJ:-OJvKjmu1Kl0IyPg6UhK:b-eaf62j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9017cf6a5adfd443d38af38fc8865b66b474d337/pkg/sql/rowexec/vector_search.go#L403)):*
> ```Go
> ) (tree.Datum, error) {
> 	randomizedVec := make(vector.T, len(queryVec))
> 	randomizedVec = v.idx.RandomizeVector(queryVec, randomizedVec)
> ```

I think randomization + quantization + encoding is something we maybe could do in a method on the wrapper index we talked about, so that it's not necessary for `rowexec` to reach down into the store.
___
*[`pkg/sql/rowexec/vector_search.go` line 483 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-OJvL1mD6ToopLWbEv-G:-OJvL1mD6ToopLWbEv-H:b-izac53) ([raw file](https://github.com/cockroachdb/cockroach/blob/9017cf6a5adfd443d38af38fc8865b66b474d337/pkg/sql/rowexec/vector_search.go#L483)):*
> ```Go
> 	ctx context.Context, flowCtx *execinfra.FlowCtx, fetchSpec *fetchpb.IndexFetchSpec,
> ) (*cspann.Index, cspann.Txn, error) {
> 	idxManager := flowCtx.Cfg.VecIndexManager.(*vecindex.Manager)
> ```

As we talked about, we can maybe create a wrapper index class that the manager returns that handles the details of txn/context.


<!-- Sent from Reviewable.io -->

DrewKimball committed:
sql: remove old debugging from fetchers

This commit removes some old debugging logic from the fetcher implementations.
These lines are not likely to be used, and complicate refactoring.

Epic: None

Release note: None

DrewKimball committed:
sql: move fetcher value decoding to helper

This commit moves the value-decoding logic for the row fetcher to be
implemented in a helper function. This will allow vector search to use
the same helper in a following commit.

Epic: CRDB-42943

Release note: None

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141897)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/rowexec/vector_search.go` line 201 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-OJvKCTQCmBFMUZb1TUD:-OK8p_PWADCj7bNXh_ds:bq6wg1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/9017cf6a5adfd443d38af38fc8865b66b474d337/pkg/sql/rowexec/vector_search.go#L201)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'm surprised there isn't a pre-existing utility function somewhere to do this. Surely it's being done elsewhere?
</blockquote></details>

I went ahead and added a helper to be used in the row fetcher and here. I left the cFetcher implementation as is for now, since it's more performance sensitive and decodes the values into a different format.
___
*[`pkg/sql/rowexec/vector_search.go` line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-OJvKjmu1Kl0IyPg6UhJ:-OK8pTY1BuwkQZnrGBE8:b-l4tc5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/9017cf6a5adfd443d38af38fc8865b66b474d337/pkg/sql/rowexec/vector_search.go#L403)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think randomization + quantization + encoding is something we maybe could do in a method on the wrapper index we talked about, so that it's not necessary for `rowexec` to reach down into the store.
</blockquote></details>

I'll leave a TODO on `QuantizeAndEncode`. Done.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
The vector index related code LGTM.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-:-OK967nN5xx-L7XOCuWR:b1hbkz7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1824 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-OK95jfR6hLIyF4bSnxM:-OK95jfR6hLIyF4bSnxN:bg5u6er) ([raw file](https://github.com/cockroachdb/cockroach/blob/6be30355a4d02490b87a5e985faca0848312bd55/pkg/sql/rowenc/index_encoding.go#L1824)):*
> ```Go
> // neededValueCols allows an early exit if all the needed columns have been
> // decoded.
> func DecodeValueBytes(
> ```

This would be an ideal function to unit-test, rather than only relying on higher level tests.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
First two commits LGTM.

> Reviewed 8 of 8 files at r3, 2 of 2 files at r4, 2 of 2 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-:-OKAW8Fm-gWcFw7Xabzz:bs99enh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @DrewKimball, and @mw5h)

___
*[`pkg/sql/rowexec/vector_search.go` line 47 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-OKAVW8hDt34zG5GzIlK:-OKAVW8hDt34zG5GzIlL:bd6ksxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6be30355a4d02490b87a5e985faca0848312bd55/pkg/sql/rowexec/vector_search.go#L47)):*
> ```Go
> }
> 
> var _ execinfra.Processor = &vectorSearchProcessor{}
> ```

nit: can combine two lines into one with `execinfra.RowSourcedProcessor`.
___
*[`pkg/sql/row/fetcher.go` line 48 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-OKATvzg7_HUiwAl5Wbw:-OKATvzg7_HUiwAl5Wbx:bxvwzjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b22b18f8f0ac9a04e984b4995571b60021a3246d/pkg/sql/row/fetcher.go#L48)):*
> ```Go
> // DebugRowFetch can be used to turn on some low-level debugging logs. We use
> // this to avoid using log.V in the hot path.
> const DebugRowFetch = false
> ```

nit: removal of `DebugRowFetch` should be done in the first commit.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-:-OKCtvtXBcwvW9xvUBpJ:bz5mair)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/row/fetcher.go` line 48 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-OKATvzg7_HUiwAl5Wbw:-OKCtoXh4VjwC0nJFBQm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b22b18f8f0ac9a04e984b4995571b60021a3246d/pkg/sql/row/fetcher.go#L48)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: removal of `DebugRowFetch` should be done in the first commit.
</blockquote></details>

Done.
___
*[`pkg/sql/rowenc/index_encoding.go` line 1824 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-OK95jfR6hLIyF4bSnxM:-OKCtShYBA8cH20NPLj0:b-30lc84) ([raw file](https://github.com/cockroachdb/cockroach/blob/6be30355a4d02490b87a5e985faca0848312bd55/pkg/sql/rowenc/index_encoding.go#L1824)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This would be an ideal function to unit-test, rather than only relying on higher level tests.
</blockquote></details>

That's a good point - I'll open up a follow-up once this one merges.
___
*[`pkg/sql/rowexec/vector_search.go` line 47 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/141897#-OKAVW8hDt34zG5GzIlK:-OKCtbxl4nBJiymULxiD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6be30355a4d02490b87a5e985faca0848312bd55/pkg/sql/rowexec/vector_search.go#L47)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: can combine two lines into one with `execinfra.RowSourcedProcessor`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
bors r=andy-kimball,yuzefovich
